aiosmtpd/smtp.py

```diff
@@ -657,19 +658,28 @@ def _strip_command_keyword(self, keyword, arg):
             return arg[keylen:].strip()
         return None
 
-    def _getaddr(self, arg):
+    def _getaddr(self, arg) -> Tuple[Optional[str], Optional[str]]:
+        """
+        Try to parse address given in SMTP command.
+
+        Returns address=None if arg can't be parsed properly (get_angle_addr /
+        get_addr_spec raised HeaderParseError)
+        """
         if not arg:
             return '', ''
-        if arg.lstrip().startswith('<'):
-            address, rest = get_angle_addr(arg)
-        else:
-            address, rest = get_addr_spec(arg)
         try:
-            address = address.addr_spec
-        except IndexError:
-            # Workaround http://bugs.python.org/issue27931
-            address = None
-        return address, rest
+            if arg.lstrip().startswith('<'):
+                address, rest = get_angle_addr(arg)
+            else:
+                address, rest = get_addr_spec(arg)
+        except HeaderParseError:
+            return None, None
+        address = address.addr_spec
+        localpart, atsign, domainpart = address.rpartition("@")
+        if len(localpart) > 64:  # RFC 5321 § 4.5.3.1.1
+            return None, None
+        else:
+            return address, rest
```

Comment:
The older exception handling method needed to workaround `bpo-27931`.

This bug has been fixed in Python >= 3.6, so the exception handling now handles `HeaderParseError` instead, which is the actual exception raised when trying to parse invalid email addresses (new test cases)